### PR TITLE
bgpv1: Adds CiliumPodIPPool Support to PathAttr Policies

### LINKED ISSUE #28296: bgpv1: Add CiliumPodIPPool Selector Type
https://github.com/cilium/cilium/pull/27705 added `AdvertisedPathAttributes` for `CiliumBGPNeighbor` with "CiliumLoadBalancerIPPool" and "PodCIDR" as supported selector types. https://github.com/cilium/cilium/pull/27100 added support for selecting and advertising [multi-pool IPAM](https://docs.cilium.io/en/stable/network/concepts/ipam/multi-pool/#ipam-crd-multi-pool) CIDRs allocated to a CiliumNode. Selector types should now be updated to support "CiliumPodIPPool".

### PR DESCRIPTION
**Pull request by danehans**
- Adds the CiliumPodIPPool selector type to BGP CP CiliumBGPPathAttributes API.
- Updates RoutePolicyReconciler to support CiliumPodIPPool.
- Added unit tests for the CiliumPodIPPool selector type.
- Regenerated CiliumBGPPeeringPolicy CRD.

Fixes: #28296

```release-note
Adds the CiliumPodIPPool selector type to BGP CP AdvertisedPathAttributes to match CiliumPodIPPool custom resources. Path attributes apply to routes announced for selected CiliumPodIPPools.
```

### CODE REVIEW

**Comment by danehans** [comment]
@rastislavs I marked this PR as a draft b/c I am not seeing the path attributes being added for prefixes being advertised from a CiliumPodIPPool. I temporarily added logging throughout for troubleshooting but I'm unable to root cause the issue. When you have time, can you take a look?

Test resources:
```
apiVersion: cilium.io/v2alpha1
kind: CiliumPodIPPool
metadata:
  creationTimestamp: "2023-09-26T21:30:52Z"
  generation: 1
  labels:
    foo: bar
  name: default
  resourceVersion: "1757"
  uid: 8c64dc70-e811-4108-b01d-f3bfa88c3a99
spec:
  ipv4:
    cidrs:
    - 10.10.0.0/16
    maskSize: 24
---
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPPeeringPolicy
metadata:
  name: cp
spec:
  nodeSelector:
    matchLabels:
      kubernetes.io/hostname: kind-control-plane
  virtualRouters:
  - exportPodCIDR: false
    localASN: 65100
    neighbors:
    - advertisedPathAttributes:
      - communities:
          standard:
          - 65001:100
        selector:
          matchLabels:
            foo: bar
        selectorType: CiliumPodIPPool
      connectRetryTimeSeconds: 120
      eBGPMultihopTTL: 1
      holdTimeSeconds: 90
      keepAliveTimeSeconds: 30
      peerASN: 65200
      peerAddress: 192.168.64.3/32
      peerPort: 179
    podIPPoolSelector:
      matchLabels:
        foo: bar
```
Logs:
[cp_not_working.log](https://github.com/cilium/cilium/files/12742437/cp_not_working.log)
[cp_working.log](https://github.com/cilium/cilium/files/12742438/cp_working.log)

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (rastislavs) [inline pkg/bgpv1/manager/route_policy_reconciler.go]: `PrefixLenMin` and `PrefixLenMax` should be equal to the actual Pool prefix length (and same for IPv6). This is most likely the reason why the policy does not apply on the advertisement.

**Comment by rastislavs** [comment]
> @rastislavs I marked this PR as a draft b/c I am not seeing the path attributes being added for prefixes being advertised from a CiliumPodIPPool. 

Thanks for the PR! I think the reason why this is not working ATM is in this comment: https://github.com/cilium/cilium/pull/28310/files#r1339661669

FYI, I plan to add a CLI command for listing the routing policies, so that similar issues are easier to debug in the future.

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (danehans) [reply pkg/bgpv1/manager/route_policy_reconciler.go]: @rastislavs thanks for the guidance. ^ was indeed the issue. Commit b9cf743 resolves this conversation.

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (rastislavs) [inline pkg/bgpv1/manager/route_policy_reconciler.go]: What about naming these `lbPoolStore` and `podPoolStore`?

Comment (danehans) [reply pkg/bgpv1/manager/route_policy_reconciler.go]: Commit 2085898 resolves this conversation.

**Comment by danehans** [comment]
cc: @harsimran-pabla @YutaroHayakawa

**Review by rastislavs — APPROVED** [review]
LGTM now, thanks!

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (harsimran-pabla) [inline pkg/bgpv1/manager/route_policy_reconciler.go]: Comment says 'advertise' but technically this is not to do any advertisement. It only creates egress policy.

File: pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go
Comment (harsimran-pabla) [inline pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go]: Do we need this constant? We do have "CiliumPodIPPool" defined [here](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2alpha1/register.go#L80 )

Comment (danehans) [reply pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go]: @harsimran-pabla thanks for the pointer. I was following the existing constant pattern but I'll update to use your ref.

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (danehans) [reply pkg/bgpv1/manager/route_policy_reconciler.go]: Thanks for catching this, I'll update.

File: pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go
Comment (danehans) [reply pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go]: @harsimran-pabla commit 51738e1 keeps the `CiliumPodIPPoolSelectorName` const but now refers to [the associated Kind](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2alpha1/register.go#L80).

Comment (rastislavs) [reply pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go]: I think it makes sense to keep this constant in the same file as the kubebuilder enum definition to make the API usage easier - referring to the associated Kind looks like a good middle ground 👍. Hopefully the kind won't ever change - as changing it would cause inconsistency with the kubebuilder enum values.

**Comment by danehans** [comment]
@harsimran-pabla when you have a moment, do you mind doing another review?

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (harsimran-pabla) [inline pkg/bgpv1/manager/route_policy_reconciler.go]: nit: should be prefixLen, same for v6

**Review by harsimran-pabla — APPROVED** [review]

File: pkg/bgpv1/manager/route_policy_reconciler.go
Comment (danehans) [reply pkg/bgpv1/manager/route_policy_reconciler.go]: @harsimran-pabla thanks for the review. Commit 677c3ca updates the var name to `prefixLen`.

**Comment by danehans** [comment]
Commit 301fa26 is a rebase to see if it fixes the `ci-l4lb` test failures.

**Comment by danehans** [comment]
@rastislavs @harsimran-pabla @YutaroHayakawa This PR is passing CI. Do you have suggestions on how to get this merged?

**Comment by YutaroHayakawa** [comment]
Seems like we still need a review from @youngnick. Once his review is in, you can put the tag `ready-to-merge` then the tophat will merge it.

**Comment by danehans** [comment]
@youngnick This PR has been hanging around for a few weeks. Do you mind reviewing when you have a moment or can you assign another reviewer?

**Comment by danehans** [comment]
@christarazi anyone else from sig-k8s you recommend to unblock this PR?

**Review by christarazi — COMMENTED** [review]
The K8s changes LGTM to me. Could you add justification in the commit msg? Without it, we lose potentially important context in git history if later down the line we bisect this commit.

**Comment by danehans** [comment]
@christarazi thanks for the review. I have updated the commit message to include additional change details and the fixed issue.

**Review by christarazi — APPROVED** [review]

**Pull request merged**
